### ISSUE TITLE
Implement `system.engine_settings`

### ISSUE THREAD

**Issue by nikitamikhaylov** [ClickHouse/ClickHouse#59913]
**Use case**

Currently we have a `system.merge_tree_settings` table which shows all the possible configuration options for the tables with MergeTree table engine. 
But there is more than on table engine in ClickHouse which also has some parameters to tune. And I would like to know them not by reading the documentation, but by selecting from the system table. 

**Describe the solution you'd like**

Add a table called `system.engine_settings` with the same structure as `system.merge_tree_settings` except one additional column - `engine_name`.

After this change we `system.merge_tree_settings` with a view with inner query `SELECT * FROM system.engine_settings WHERE engine_name='MergeTree'`

**Additional context**

We can extend the `IStorage` interface and add a method to return a list of settings. Then we can add a few methods inside the `StorageFactory` to return settings for a specified engine name.

**Comment by Blargian** [comment]
@nikitamikhaylov I'd like to work on this

**Comment by nikitamikhaylov** [comment]
@Blargian Thank you!

**Comment by Ergus** [comment]
Hi @nikitamikhaylov 

I am already working on this. So far the task is easy, but there are several sparse modifications needed because, unlike `MergeTrees`, the other engines do not contain any set of valid options we can iterate over, but they are declared in a namespace as `extern` instances i.e:

https://github.com/ClickHouse/ClickHouse/blob/c3ff0f61ed381415978f3e05e7518cad65bfef94/src/Storages/StorageFile.cpp#L86-L92

The first step require to use the macros in: 

https://github.com/ClickHouse/ClickHouse/blob/c3ff0f61ed381415978f3e05e7518cad65bfef94/src/Core/BaseSettings.h#L889

For all the other engines in order to avoid duplication in the latter implementation. This will provide an iterable structure equivalent to:

https://github.com/ClickHouse/ClickHouse/blob/c3ff0f61ed381415978f3e05e7518cad65bfef94/src/Storages/MergeTree/MergeTreeSettings.h#L54

but for every engine.

I think this is somehow the right thing to do in order to have a more uniform implementations of all our storage engines, but somehow I find it a bit overkill if we are only interested in this feature.

If anyone has a better (simpler) alternative in mind, please feel free to suggest it.

**Comment by Algunenano** [comment]
>  the other engines do not contain any set of valid options we can iterate over,

Most other settings will have an internal BaseSettings::changes() you can use and expose (like we expose it for MergeTreeSettings).  I expect the case where an engine is doing custom things like StorageFile to be a corner case.

**Comment by Ergus** [comment]
Hi @Algunenano 

The problem is not to access internal settings information, but to get the list of settings for every engine. In the current implementation there is not clear association between them, only the forward declaration at the beginning, but "formally" it is like any engine can use any setting.

It looks like, at some point, the intended method to declare settings was to use the family of macros like `DECLARE_SETTINGS_TRAITS`. But, I don't know why,  I don't see any engine using this approach. I can't say why, probably because many settings are repeated (or shared) by many engines.

Only `MergeTrees` seems to do that with a multiple level of wrappers and inheritances overly complicated: `MergeTreeSettings` enclosing `MergeTreeSettingsImpl` that inherits from `BaseSettings<MergeTreeSettingsTraits>`, where  `MergeTreeSettingsTraits` is created using `DECLARE_SETTINGS_TRAITS`.

In order to allow other engines to produce a list of settings that we can "process" to create a table, the first step would be to modify all other engines use `DECLARE_SETTINGS_TRAITS`, probably creating some BaseSettings<MyEngineSettingsTraits>.

Only after that we will be allowed to construct the table columns.

**Comment by Algunenano** [comment]
> It looks like, at some point, the intended method to declare settings was to use the family of macros like DECLARE_SETTINGS_TRAITS

This is exactly how it works now. If you look for `DECLARE_SETTINGS_TRAITS` you'll find 20+ different setting types, e.g. MemorySettings, HiveSettings, etc. Just as MergeTreeSettings.

> Only MergeTrees seems to do that with a multiple level of wrappers and inheritances overly complicated: MergeTreeSettings enclosing MergeTreeSettingsImpl that inherits from BaseSettings<MergeTreeSettingsTraits>, where MergeTreeSettingsTraits is created using DECLARE_SETTINGS_TRAITS.

Yeah, we used to not have `*Impl` and it was a nightmare. Every minor change to a setting (even a comma in a description) required rebuilding 3000 files. It also made the resulting binary problematic in multiple places due to size. That's why now it's private.

**Comment by kaijaytu** [comment]
Hi @nikitamikhaylov, I noticed this issue was recently unassigned. I'd like to take it up if it's still available.

I've read through the discussion between @Ergus and @Algunenano about the implementation approach -- using DECLARE_SETTINGS_TRAITS macros to make engine settings iterable, then building the system.engine_settings table on top of that. I'm happy to continue from where this was left off.

Could you assign this to me?

**Comment by Ergus** [comment]
Hi @kaijaytu 

Very thanks for taking this over. I haven't have time to put this in the head of priority stack, but if you need any help at some point, do not hesitate to request or mention me. Also ask me for review when you open the PR.

**Comment by kaijaytu** [comment]
Hi @Ergus, I'm working on this issue. Quick question about PR conventions: does ClickHouse have a preferred PR title format? I noticed most commits use plain imperative sentences without issue references in the title.

Would a JIRA-style prefix like [#59913] Implement system.engine_settings table be acceptable, or should I stick with just Implement system.engine_settings table and rely on the PR description for the issue link?

**Comment by buggins** [comment]
Taking this forward alongside #116543, at @nikitamikhaylov's suggestion: https://github.com/ClickHouse/ClickHouse/pull/118544. @kaijaytu's
commits from #113269 are on my branch with their authorship intact; details in https://github.com/ClickHouse/ClickHouse/pull/113269#issuecomment-5569852858.

`system.engine_settings` now reports the settings an engine actually uses on this server rather than
compiled defaults, so the view this issue asks for is exactly expressible. Measured rather than
assumed, against `system.merge_tree_settings`:

- **Names** — a two-way `EXCEPT` returns 0 rows on either side. Same for
  `system.replicated_merge_tree_settings`.
- **Values** — identical under a `<merge_tree>` config override, and under `--compatibility=23.3`
  (48 changed on both sides). 0 disagreeing values on the replicated pair.
- **Columns** — identical, and in the same order, but for the leading `engine_name`.

So the view would be a plain projection with `WHERE engine_name = 'MergeTree'`.

To be clear about status: that pull request is a test task submitted with a job application, so it
may not be accepted. #113269 stays open and remains the standing proposal for this issue
independently of what happens to mine.

**I have kept the replacement itself out of scope.** The issue proposes it, but it changes a table a
lot of tooling reads, and the safe default is that `system.merge_tree_settings` keeps its present
behaviour exactly. If the replacement is wanted later, one detail is worth recording now: should
`system.engine_settings` gain alias rows (an open question in the new PR, since `system.settings`
has them and `merge_tree_settings` does not), the view can carry `AND alias_for = ''` and still
reproduce today's output row for row. The two questions are therefore independent — neither has to
wait on the other.
